l1-contracts/contracts/bridge/asset-router/L2AssetRouter.sol

```diff
@@ -151,11 +152,15 @@ contract L2AssetRouter is AssetRouterBase, IL2AssetRouter {
             _data: _assetData
         });
 
-        bytes memory message = _getL1WithdrawMessage(_assetId, _l1bridgeMintData);
-        if (L2_LEGACY_SHARED_BRIDGE != address(0)) {
+        bytes memory message;
+        if (L2_LEGACY_SHARED_BRIDGE == address(0)) {
+            message = _getAssetRouterWithdrawMessage(_assetId, _l1bridgeMintData);
             // slither-disable-next-line unused-return
             L2ContractHelper.sendMessageToL1(message);
         } else {
+            address l1Token = IL2NativeTokenVault(L2_NATIVE_TOKEN_VAULT_ADDR).tokenAddress(_assetId);
+            (uint256 amount, address l1Receiver) = abi.decode(_assetData, (uint256, address));
```

Comment:
This is not compatible with using custom asset handlers

Reply:
Actually, the most imcompatible part is here:

```
            address l1Token = IL2NativeTokenVault(L2_NATIVE_TOKEN_VAULT_ADDR).tokenAddress(_assetId);
```

we always demand the asset to go through the native token vault

Reply:
Fixed